sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala

```diff
@@ -795,7 +795,7 @@ class FlatMapGroupsWithStateSuite extends StateStoreMetricsTest {
         }
       },
       CheckNewAnswer(("c", "-1")),
-      assertNumStateRows(total = 0, updated = 0)
+      assertNumStateRows(total = 0, updated = 1, lateInput = 0)
```

Comment:
This is a bug on FlatMapGroupsWithStateExec: when state operators drop rows from state, these operators don't increase "updated". Unfortunately FlatMapGroupsWithStateExec does. I'll raise a patch shortly.

Reply:
Maybe that is intention to increase the value, but the "updated" value is override with 0 when exposing if the batch was run with empty data (that's why the value was "0" though the row was removed). So the value is applied conditionally which is confusing. I'd rather say we need to add "removed" metrics or just follow the approach other state operators are doing.

Reply:
> The "updated" value is override with 0 when exposing if the batch was run with empty data

This can be considered as a bug instead, as we have actual case where batch with empty data has nonzero "updated" value. I guess it's added when batch with empty data wasn't introduced, and missed to address the case. I'll deal with this as well. (The fix is actually in this PR, but given the issue is orthogonal to the origin issue of this PR, so better to have new PR.)

Reply:
SPARK-29312 for FlatMapGroupsWithStateExec behavior fix (have a patch, but maybe I'd just keep it as is if it's turned out to be intended), and SPARK-29314 for ProgressReporter fix (#25987)

Reply:
#25987 is now merged in.